rust/bin-shared/src/tun_device_manager/windows.rs

```diff
@@ -274,36 +292,76 @@ impl tun::Tun for Tun {
     }
 
     /// Check if more packets can be sent.
-    fn poll_send_ready(&mut self, _: &mut Context) -> Poll<io::Result<()>> {
-        // TODO: Figure out how we can do readiness checks on `wintun`.
+    fn poll_send_ready(&mut self, cx: &mut Context) -> Poll<io::Result<()>> {
+        ready!(self
+            .outbound_tx
+            .poll_reserve(cx)
+            .map_err(io::Error::other)?);
 
         Poll::Ready(Ok(()))
     }
 
     /// Send a packet.
     fn send(&mut self, packet: IpPacket) -> io::Result<()> {
-        let bytes = packet.packet();
-
-        let len = bytes
-            .len()
-            .try_into()
-            .map_err(|_| io::Error::other("Packet too large; length does not fit into u16"))?;
-
-        let mut pkt = self
-            .session
-            .allocate_send_packet(len)
+        self.outbound_tx
+            .send_item(packet)
             .map_err(io::Error::other)?;
 
-        pkt.bytes_mut().copy_from_slice(bytes);
-        // `send_packet` cannot fail to enqueue the packet, since we already allocated
-        // space in the ring buffer.
-        self.session.send_packet(pkt);
-
         Ok(())
     }
 }
 
-// Moves packets from the user towards the Internet
+// Moves packets from Internet towards the user
+fn start_send_thread(
+    mut packet_rx: mpsc::Receiver<IpPacket>,
+    session: Arc<wintun::Session>,
+) -> io::Result<std::thread::JoinHandle<()>> {
+    // See <https://learn.microsoft.com/en-us/windows/win32/debug/system-error-codes--0-499->.
+    const ERROR_BUFFER_OVERFLOW: i32 = 0x6F;
+
+    std::thread::Builder::new()
+        .name("TUN send".into())
+        .spawn(move || {
+            'recv: loop {
+                let Some(packet) = packet_rx.blocking_recv() else {
+                    tracing::info!(
+                        "Stopping TUN send worker thread because the packet channel closed"
+                    );
+                    break;
+                };
+
+                let bytes = packet.packet();
+
+                let Ok(len) = bytes.len().try_into() else {
+                    tracing::warn!("Packet too large; length does not fit into u16");
+                    continue;
+                };
+
+                let mut pkt = loop {
+                    match session.allocate_send_packet(len) {
+                        Ok(pkt) => break pkt,
+                        Err(wintun::Error::Io(e))
+                            if e.raw_os_error()
+                                .is_some_and(|code| code == ERROR_BUFFER_OVERFLOW) =>
+                        {
+                            tracing::debug!("WinTUN ring buffer is full");
+                            std::thread::sleep(Duration::from_millis(10)); // Suspend for a bit to avoid busy-looping.
+                        }
+                        Err(e) => {
+                            tracing::error!("Failed to allocate WinTUN packet: {e}");
```

Comment:
My hypothesis is that we should actually be shutting everything down if this happens. By logging it on ERROR, we'll see it in Sentry if this happens in the wild.

Reply:
Is it possible we'll also hit this error path if the adapter is shutting down when we try to send a packet?

https://github.com/WireGuard/wintun/blob/master/api/wintun.h#L245

Reply:
Yeah that is what I meant above.

We never initiate a shutdown other then when dropping `Tun` so this shouldn't ever happen (unless the adapter is force-removed for example)